ibtk/include/ibtk/HierarchyAveragedDataManager.h

```diff
@@ -0,0 +1,251 @@
+     * deviation.
+     */
+    inline bool updateTimeAveragedSnapshot(int u_idx,
+                                           double time,
+                                           SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                           const int wgt_idx = IBTK::invalid_index,
+                                           double tol = 1.0e-8)
+    {
+        return updateTimeAveragedSnapshot(u_idx, time, hierarchy, d_mean_refine_type, wgt_idx, tol);
+    }
+    bool updateTimeAveragedSnapshot(int u_idx,
+                                    double time,
+                                    SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy,
+                                    const std::string& mean_refine_type,
+                                    const int wgt_idx = IBTK::invalid_index,
+                                    double tol = 1.0e-8);
+
+    /*!
+     * Return true if all the tracked mean fields are at a steady state.
+     */
+    inline bool isAtPeriodicSteadyState()
+    {
+        bool steady_state = true;
+        for (const auto& idx_steady_state : d_idx_steady_state_map)
+        {
+            steady_state = steady_state && idx_steady_state.second;
+        }
+        return steady_state;
+    }
+
+    /*!
+     * Return whether the point at the specified time is at a periodic steady state.
+     */
+    inline bool isAtPeriodicSteadyState(double time, const double tol)
+    {
+        time = getTimePt(time, tol);
+        return d_idx_steady_state_map.at(time);
+    }
+
+    /*!
+     * Return the time points this class is storing.
+     */
+    inline const std::set<double>& getSnapshotTimePts()
+    {
+        return d_snapshot_time_pts;
+    }
+
+    /*!
+     * Returns the exact data point at which points are being stored in the internal map. Note that if a time point is
+     * not found within the specified tolerance, an unrecoverable error occurs.
+     */
+    double getTimePt(double time, double tol);
+
+    /*!
+     * Get the SnapshotCache object
+     */
+    const std::unique_ptr<SnapshotCache>& getSnapshotCache()
+    {
+        return d_snapshot_cache;
+    }
+
+    /*!
+     * Set the threshold for achieving a steady state.
+     */
+    void setSteadyStateThreshold(const double threshold)
+    {
+        d_periodic_thresh = threshold;
+    }
+
+private:
+    /*!
+     * \brief Registers a scratch variable with the variable database. This data is allocated and deallocated as needed.
+     */
+    void commonConstructor(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
+                           SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > hierarchy);
+    /*!
+     * \brief Default constructor.
+     *
+     * \note This constructor is not implemented and should not be used.
+     */
+    HierarchyAveragedDataManager() = delete;
```

Comment:
We can do this, but
1. Since we have a non-default constructor the default one is implicitly deleted
2. Since we have non-copyable members (the unique pointer) the coppy and assignment constructors are implicitly deleted

so we don't need these.